### ISSUE TITLE
json: cannot unmarshal array into Go value of type types.NetworkResource in 17.12.0 RC1, but not under 17.11

### ISSUE THREAD

**Issue by denefoster** [moby/moby#35755]
**Description**

docker stack deploy with an external network under 17.12.0 rc1 results in a Go error message:

```json: cannot unmarshal array into Go value of type types.NetworkResource```

**Steps to reproduce the issue:**
1. Create a docker compose file with an external network
```
version: "3.3"

services:
  node-exporter:
    hostname: "{{.Node.ID}}-node-exporter"
#    image: prom/node-exporter
    image: reg.slush.ca:5000/node-exporter
    deploy:
      mode: global
    environment:
      - "NODE_ID={{.Node.ID}}"
    volumes:
      - /proc:/host/proc:ro
      - /sys:/host/sys:ro
      - /:/rootfs:ro
      - /etc/hostname:/etc/nodename
    command:
      - '-collector.textfile.directory=/etc/node-exporter/'
      - '-collector.procfs=/host/proc'
      - '-collector.sysfs=/host/sys'
      - '-collector.filesystem.ignored-mount-points="^(/rootfs|/host|)/(sys|proc|dev|host|etc)($$|/)" collector.filesystem.ignored-fs-types="^(sys|proc|auto|cgroup|devpts|ns|au|fuse\.lxc|mqueue)(fs|)$$"'
      - '-collectors.enabled=textfile,conntrack,diskstats,entropy,filefd,filesystem,loadavg,mdadm,meminfo,netdev,netstat,stat,time,vmstat'
    ports:
      - target: 9100
        published: 9100
        mode: host
    networks:
      - node

networks:
  node:
    external:
      name: host
```
2. docker stack deploy -c file.yml node

**Describe the results you received:**

```
dene@legion ~/services/node_exporter𝛀 docker stack deploy -c docker-compose.yml node-exporter
json: cannot unmarshal array into Go value of type types.NetworkResource
```
**Describe the results you expected:**

```
dene@taylor-swift ~𝛀 docker stack deploy -c file.yml node
Creating service node_exporter_node-exporter
```

This works under:
```
Client:
 Version:      17.11.0-ce
 API version:  1.34
 Go version:   go1.8.4
 Git commit:   1caf76c
 Built:        Mon Nov 20 18:30:18 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.12.0-ce-rc1
 API version:  1.35 (minimum version 1.12)
 Go version:   go1.9.2
 Git commit:   ee2f943
 Built:        Thu Dec  7 17:25:28 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:	17.12.0-ce-rc1
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	ee2f943
 Built:	Thu Dec  7 17:26:53 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce-rc1
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	ee2f943
  Built:	Thu Dec  7 17:25:28 2017
  OS/Arch:	linux/amd64
  Experimental:	true
```

**Output of `docker info`:**

```
Containers: 1531
 Running: 6
 Paused: 0
 Stopped: 1525
Images: 26
Server Version: 17.12.0-ce-rc1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3228
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: vn20ge15x4wi36973pn7nkvdp
 Is Manager: true
 ClusterID: mi7qu4jkfxivybqp4phd8uv5r
 Managers: 3
 Nodes: 10
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 192.168.1.251
 Manager Addresses:
  192.168.1.220:2377
  192.168.1.230:2377
  192.168.1.251:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729 (expected: v1.0.0)
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-103-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 15.67GiB
Name: legion
ID: DZQX:TT4I:GSXU:KWJG:SPUO:FXEN:M5XJ:OGKW:GGX7:I6FD:SDG6:KXMY
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by cpuguy83** [comment]
ping @andrewhsu @dnephin

**Comment by djeeg** [comment]
Getting the same with external networks (stack managed networks seem to be okay)

```
version: '3.5'

# docker network create --attachable -d overlay nettmp
# docker network rm nettmp

# docker stack deploy --compose-file tmp.yml stacktmp
# docker stack services stacktmp
# docker stack ps stacktmp
# docker stack rm stacktmp

# docker exec -it $(sudo docker ps | grep "nginx" | awk '{print $1}') /bin/bash

services:

  nginx:
    image: nginx
    networks:
     nettmp:

networks:

  #external managed network
  nettmp:
    name: "nettmp"
    external: true

  #stack managed network
  #nettmp:
```

**Comment by djeeg** [comment]
Thanks, looks fixed in 17.12.0-ce-rc3
